.palantir/revapi.yml

```diff
@@ -261,6 +261,16 @@ acceptedBreaks:
         \ T) throws java.io.IOException, com.fasterxml.jackson.core.JacksonException\
         \ @ org.apache.iceberg.rest.RESTSerializers.UpdateRequirementDeserializer"
       justification: "False positive - JacksonException is a subclass of IOException"
+    - code: "java.method.numberOfParametersChanged"
+      old: "method org.apache.iceberg.io.FileAppender<org.apache.iceberg.ManifestEntry<F>>\
+        \ org.apache.iceberg.ManifestWriter<F extends org.apache.iceberg.ContentFile<F\
+        \ extends org.apache.iceberg.ContentFile<F>>>::newAppender(org.apache.iceberg.PartitionSpec,\
+        \ org.apache.iceberg.io.OutputFile)"
+      new: "method org.apache.iceberg.io.FileAppender<org.apache.iceberg.ManifestEntry<F>>\
+        \ org.apache.iceberg.ManifestWriter<F extends org.apache.iceberg.ContentFile<F\
+        \ extends org.apache.iceberg.ContentFile<F>>>::newAppender(org.apache.iceberg.PartitionSpec,\
+        \ org.apache.iceberg.io.OutputFile, java.lang.String, java.lang.Integer)"
```

Comment:
rather than having to break the API I think it would be better to define a new `newAppender(..)` method that by default calls the original one with null/default params. See also my comment further below.

Once this is done, you shouldn't have any diff on `revapi.yml` anymore.

Reply:
I think you have it backwards, the original `newAppender` has 2 parameters, and we need a new `newAppender` with 4 parameters. The `newAppender` with 2 parameters can call the one with 4 parameters, passing null for the additional parameters, and we can make this the default for the `newAppender` with 2 parameters, but then it is no longer abstract, and we still have the same revapi breakage. (We get a "java.method.numberOfParametersChanged" for the abstract `newAppender` method.)
I had considered and tried this, and rejected it for that reason.

Reply:
I think we can still do it without breaking the API. I created https://github.com/apache/iceberg/commit/1b237d570197d52fc0c7d959abc28ff14830d9aa in my fork that outlines it. 

Basically we need to go through one deprecation cycle (one minor release) as indicated in https://github.com/apache/iceberg/blob/master/CONTRIBUTING.md#minor-version-deprecations-required before we can break the API here.

Reply:
Retained but deprecated the old `newAppender` method, but adding the new one still causes a revapi breakage, as I mentioned before.

Reply:
this is because the new method is also `abstract`. If you make it non-abstract (like https://github.com/apache/iceberg/commit/667aacc59ed5d8b70baf94c6c290f7538d00f790#diff-4e8faf21494a38d340335de99357de31cdca2f3f82af495f832f67fc2446ad49R79), then it won't break the API

Reply:
I fail to see how that helps:
```
  protected FileAppender<ManifestEntry<F>> newAppender(
      PartitionSpec spec,
      OutputFile outputFile,
      String compressionCodec,
      Integer compressionLevel) {
    return newAppender(spec, outputFile);
  }
```
We want to remove the `newAppender` with two parameters in the future release, at which time the above has to be changed. We really want to make it abstract. At that point, we will have a revapi breakage, wouldn't we? I fail to see how we can avoid the revapi breakage.
@rdblue @danielcweeks ?

Reply:
I looked at some of the historical justifications in the revapi.yml. I see, e.g., a number of
```
      justification: "Allow adding a new method to the interface - old method is deprecated"
```
I think that is the way, isn't it? Deprecate the old abstract `newAppender` method (with 2 parameters), add a new abstract `newAppender` method (with 4 parameters), and justify it as
```
      justification: "Allow adding a new method to the abstract class - old method is deprecated"
```
Otherwise, how would you ever change the API? Am I missing something?

Reply:
Here is an example: https://github.com/apache/iceberg/commit/128d5a161fda076118a7cab1d95ab5064400e08a.
Deprecate old method and add new replacement method at the same time.

Reply:
the example you mentioned is quite old and this was before we were stricter on API breakages. 
As part of #6146 we went back and fixed API breakages by first deprecating them first. In #6274 we then went ahead and removed those deprecations.

Reply:
I probably don't understand how revapi works.
Thank you for the examples, but unfortunately they do not answer the question on my mind. In your example, in the first change, you deprecate a number of methods and annotate that a replacement method should be used. In those cases, the replacement method had already been added (by some earlier change). Then in the second change, the deprecated methods are removed.
I completely understand deprecating a method in one release, and removing it in the following release. What I don't understand is how to add the replacement method, or how to add a new method.
In your example, in the first change, either the replacement method has already been added by some earlier change, or you amend a new method to have a default implementation (e.g., throwing an `UnsupportedOperationException`). In the example, this occurs in interfaces. There is unfortunately no example of an abstract class. If I'm not mistaken, you can add a default implementation to a method in an interface, but you cannot add a default implementation to an abstract method in an abstract class, as the method is then no longer abstract.

Let's say I do as you suggest and make the `newAppender` with 4 parameters non-abstract as follow:
```
  protected FileAppender<ManifestEntry<F>> newAppender(
      PartitionSpec spec,
      OutputFile outputFile,
      String compressionCodec,
      Integer compressionLevel) {
    return newAppender(spec, outputFile);
  }
```
In the following release, I'd remove the deprecated old `newAppender` with 2 parameters (well and good), and I'd need to change the `newAppender` with 4 parameters to be abstract, since the default implementation no longer makes sense. Wouldn't that break revapi at that point?

Anyway, I just want to get this change in, and I make the change you suggest, but it doesn't seem right to me.

Reply:
https://github.com/apache/iceberg/blob/master/CONTRIBUTING.md#minor-version-deprecations-required talks about deprecating things to be removed, but not about adding things. What would help is documentation about how to add new things.